Test CLI key validation and SM2 key validation

romen opened the pull request:
This PR adds a new series of tests to test `EVP_PKEY` objects validation thorugh the `openssl pkey -check` CLI.

- test corner cases for P-256 private keys as an example for EC keys
- `openssl pkey -check` did not return failure if the check fails: instead of parsing the output we fix `pkey` to return an error in this case
- add test for a corner case for SM2 private keys from #8435 
- create dedicated validate function for SM2 keys to check their validity to partially fix one of the issues highlighted in #8435

Partially fixes #8435

##### Checklist
- [x] tests are added or updated

slontis commented on apps/pkey.c:
```diff
@@ -258,6 +258,8 @@ int pkey_main(int argc, char **argv)
                            ERR_reason_error_string(err));
                 ERR_get_error(); /* remove err from error stack */
             }
+            EVP_PKEY_CTX_free(ctx);
```
It is arguably cleaner to move the scope of ctx to the top level.. and then just do the free after label end:

romen replied:
Done: 24f10b9b611067ce62342f2c6fa5d877a669519b

slontis commented on crypto/ec/ec_key.c:
```diff
@@ -517,6 +519,34 @@ int ec_key_private_check(const EC_KEY *eckey)
+        return 0;
+    }
+
+    /* range of SM2 private key is [1, n-1) */
```
]?

romen replied:
This is intentional, `1` is included in the range, `n-1` is excluded from the range.

slontis commented on test/recipes/91-test_pkey_check.t:
```diff
@@ -0,0 +1,58 @@
+my @tests = ();
+
+push(@tests, qw(
+    ec_p256_bad_0.pem
```
Maybe comment on what is actually 'bad' for each of these files.

romen replied:
See bfaaae1a47a99dcdd8bb947fb7d991cce27446e5 and 6efa12eb6f1305d6f78ac6f5096da22aa78221e6

slontis reviewed (commented):
Apart from some minor issues - this looks ok to me.

t-j-h commented:
I would separate out the apps change into a separate PR as it makes sense independent of the SM2 changes.
Basically all error conditions except the "check" will exit with 1, only check which fails does not.

romen commented:
> I would separate out the apps change into a separate PR as it makes sense independent of the SM2 changes.
> Basically all error conditions except the "check" will exit with 1, only check which fails does not.

Having a separate PR that does it for all corner cases and across all the apps (and adds the tests to make sure we are doing the right thing in all cases) is a massive effort that I cannot put in right now.
Separating only a PR that does the breaking change for `pkey -check` or `pkey -pubcheck` could be an alternative, to discuss the breaking change, but that PR to be complete would require tests, and this is the PR that includes the required tests.

---

I could have this PR only introduce the `pkey -[pub]check` behavior change and the non-SM2 tests, creating a separate PR to address (part of) the SM2 issue from #8435 with the SM2 tests and the SM2 fix, depending on this PR.
The reasons I see against doing that are:
- the behavior change is motivated by the SM2 issue, without it we would be adding a test for something that is already passing in 1.1.1 and 3.0.0
- test coverage for the behavior change would be incomplete: testing the SM2 issue is how in the PR we can test that keys that should not pass validation are passing validation

romen commented:
I plan to open a separate OTC vote to approve the change in behavior. See https://www.mail-archive.com/openssl-project@openssl.org/msg02241.html for the vote proposal.

I am applying the OTC hold label until the vote is resolved.

romen commented:
While the OTC process makes its progress, we can still review the overall PR so that by the time a decision is taken the PR would be ready to be merged.

@slontis I should have addressed all the feedback so far, and I plan to shortly add a commit updating CHANGES highlighting the change in behavior of `pkey`.

romen commented:
- See 74045730b2 for a `fixup!` commit adding an entry in `CHANGES.md` for the breaking behavior change for `apps/pkey`
- See 0276134f94 for a `fixup!` commit adding an entry in `CHANGES.md` for the improved SM2 key validation

romen commented on test/recipes/91-test_pkey_check.t:
```diff
@@ -0,0 +1,62 @@
+        ok(!check_key($f), $str);
+    } else {
+        diag("Missing file $f");
+        ok(0, $str);
```
@levitte , stylewise is there a better way to write this?
I saw https://perldoc.perl.org/Test::More#fail would it be more appropriate?

levitte replied:
One of these:

```
    if (-s $f) {
        ok(!check_key($f), $str);
    } else {
        fail("Missing file $f");
    }
```

or

``` perl
    SKIP: {
        skip "Missing file $f", 1 unless -s $f;
        ok(!check_key($f), $str);
    }
```

Which variant you choose depends on what you think this recipe is meant to test.  Is it meant to test the presence of those files?  In that case the former.  Is it only meant to test that `openssl pkey` fails for certain keys?  Then the latter seems more appropriate.

romen replied:
This is tricky!!

I conceptually and stylistically like the second option with the `SKIP` block... but, what I don't like is that I would consider the absence (because it was removed or renamed) of one of the files explicitly mentioned in the test recipe as a severe error in the recipe, and that is why functionally I would prefer the first solution........

Can you help me decide among the two options? You also have a better perspective on what most other recipes do about similar cases, so I would go with the option you prefer here.

levitte replied:
The only similar test that I can think of that does the same kind of check is the fuzz test.  There, we have explicitly allowed the corpora to be absent.  From your description, this one is different and you do want to have a check that the files are indeed present, and since you otherwise check that the program invocation fails (which is the actual unusual bit), that check must be explicit in this recipe.

With all that in mind, the first variant (the one with the `fail` call) seems to be the most appropriate.

romen replied:
Addressed with 05bd0080fc1cc59a10a6a7b9b1818c763ebec1a5

romen commented:
> While the OTC process makes its progress, we can still review the overall PR so that by the time a decision is taken the PR would be ready to be merged.

The vote I proposed should really be an OMC vote about allowing the breaking change (either for the specific case in question or more in general for this kind of situations). I won't be opening the OTC vote as I planned, and will wait for an OMC resolution.

levitte commented on crypto/ec/ec_key.c:
```diff
@@ -517,6 +519,34 @@ int ec_key_private_check(const EC_KEY *eckey)
 }
 
+#ifndef OPENSSL_NO_SM2
+int sm2_key_private_check(const EC_KEY *eckey)
```
This feels hugely misplaced.  Create a new file in `crypto/sm2`, `sm2_key.c` perhaps?

romen replied:
That was my first idea as well, but then I realized that SM2 keygen is still happening in this file and it felt wrong to put validation somewhere else!

Would you insist on splitting validation in its own file?

levitte replied:
This is specifically an *sm2* function...  does that clarify how I think?  The key generation is combined in one function that looks at the internal key type (or more precisely, the flag `EC_FLAG_SM2_RANGE`) as indication how it should generate the key.

You *could* choose to include the SM2 validation in `ec_key_private_check()`.  That would be much more acceptable in this file, in my view.

Choose.

romen replied:
I compromised with an `sm2_key.c` file, but in the `crypto/ec` folder, as we don't have low level SM2 objects so `sm2_key_private_check()` feels to me like an `EC_KEY` operation (that is what it operates on).

See 30ba1329ed8d6d11c1ad75728f50b87b2f38aba2

levitte replied:
Er, files in `crypto/sm2/` operate on `EC_KEY` objects as well, but `crypto/sm2/` is already consecrated SM2 specific functions (in terms of *functionality*, not in terms of types)...  having one such file (and function) in `crypto/ec/` breaks the existing file organization.  It comes off as a bit jarring, related to already existing stuff.

If that makes you feel better, everything in `crypto/sm2/` could be moved to `crypto/ec/`, since we've already piled everything else that's EC related there (such as Ed25519 et al).

romen replied:
You got me! I completely erased from my mind that a `crypto/sm2` folder already existed and is populated with similar SM2 specific uses of `EC_KEY` objects, I agree with you and will move it in a upcoming PR!

Thanks for your insistence!!

romen replied:
See 2d233c372251f541c119210a1c3c136d83454720

I am not particularly proud of

```C
#include "../ec/ec_local.h"
```

but I feel internal resistance to the idea of rewriting it to use the getters: if `SM2_KEY` existed we would do direct access in this kind of low level function, as we do in the `ec_key_private_check()` function from which the SM2 specific variant is derived.

So unless @levitte or another reviewer feels strongly about avoiding this practice for this specific case, I would keep it.
My rationale, even if in general I dislike the practice of including internal headers from other submodules, is that the `crypto/sm2` case is special in being a _conceptual_ grouping (which is intentionally not matched with dedicated data types for the submodule), and future functions belonging in `sm2_key.c` would likely share with `sm2_key_private_check()` the need to access the fields of `EC_KEY` data objects at the lowest level of abstraction.

levitte replied:
I would like to see you use the getters, just like the rest of the `crypto/sm2/` code.

romen replied:
Addressed with b5b1407ff7650db5641a0b8855177f3ef6df413a

romen commented:
I noticed the conflict on `CHANGES.md`: I plan to wait to do a rebase and force-push until the outstanding conversations are resolved, to ease their incremental review:

- https://github.com/openssl/openssl/pull/13359#discussion_r521685617
- https://github.com/openssl/openssl/pull/13359#discussion_r521700660

Locally the tests are passing with and without `no-sm2`, and in any case I think there will be enough time for CI to run after the force-push and before the OMC starts and closes a vote on the breaking change aspect.

romen commented:
@levitte please ping me if you consider both conversation resolved, so I will proceed with the rebase and force push to address the `CHANGES` conflict.

romen commented:
> @levitte please ping me if you consider both conversation resolved, so I will proceed with the rebase and force push to address the `CHANGES` conflict.

ping @levitte

romen commented:
> > While the OTC process makes its progress, we can still review the overall PR so that by the time a decision is taken the PR would be ready to be merged.
> 
> The vote I proposed should really be an OMC vote about allowing the breaking change (either for the specific case in question or more in general for this kind of situations). I won't be opening the OTC vote as I planned, and will wait for an OMC resolution.

Ping @openssl/omc for opening a vote on the behavior change for this specific case (or something more general as some OMC members discussed on openss-project).

levitte commented:
> @levitte please ping me if you consider both conversation resolved, so I will proceed with the rebase and force push to address the `CHANGES` conflict.

Ah, different ways of doing things.  I often resolve when I think I've fixed a review comment.  If there are still issues, I expect that people will start new reviews with new change requests.

Anyway, I'm having a look.

levitte commented on apps/pkey.c:
```diff
@@ -258,8 +258,8 @@ int pkey_main(int argc, char **argv)
             }
+            goto end;
         }
-        EVP_PKEY_CTX_free(ctx);
```
You could have moved this line to appear before `if (r == 1) {` (line 245).

romen replied:
It's not the best for debugging, as if one wants to inspect the ctx in case of error it has already been freed.

I can of course comply with your preference if you think it would make things more readable (I imagine this would also mean putting back the declaration of ctx inside this block rather than outside).

levitte replied:
Er, hmm?  Oh, yeah, ok, I can understand debugability.  That's actually sufficient explanation for the modification you made, thank you.

levitte commented:
Please rebase so we can see CI results.  Other than that, I have no further comment for the moment

romen commented:
Force pushed.

There are now link error for `providers/fips.so` (by the way they are being detected only at runtime when DSO loads it, and not at build time).

The problem is that when `fips.so` is loaded there is a missing `sm2_key_private_check` symbol.

This happens because `providers/implementations/keymgmt/ec_kmgmt.c` has a reference to it (correctly guarded by `OPENSSL_NO_SM2`) but as documented in `providers/build.info`, `libimplementations.a` is built once for both the fips module and the default module.

I read the documentation and its suggested workarounds (based on splitting source files and adding them as sources for `libnonfips.a`), but I don't know how to reconcile them with the fact that we want the SM2 and the EC kmgmt to be in the same file to avoid code duplication.


@levitte suggestions?

levitte commented:
> This happens because `providers/implementations/keymgmt/ec_kmgmt.c` has a reference to it (correctly guarded by `OPENSSL_NO_SM2`) but as documented in `providers/build.info`, `libimplementations.a` is built once for both the fips module and the default module.

Wait what?????  Please point at the comment that's so badly written, that needs fixing!

> I read the documentation and its suggested workarounds (based on splitting source files and adding them as sources for `libnonfips.a`), but I don't know how to reconcile them with the fact that we want the SM2 and the EC kmgmt to be in the same file to avoid code duplication.

I'll comment in the diff

levitte commented on crypto/sm2/build.info:
```diff
@@ -1,5 +1,5 @@
 LIBS=../../libcrypto
 SOURCE[../../libcrypto]=\
-        sm2_sign.c sm2_crypt.c sm2_err.c
+        sm2_sign.c sm2_crypt.c sm2_err.c sm2_key.c
```
It's actually strange that none other is needed that way, but anyway, to get `sm2_key.c` included with the FIPS module as well, you must add this:

```
SOURCE[../../providers/libfips.a]=sm2_key.c
```

romen replied:
Uhm... But why should there be SM2 code in the fips specific dependencies? It's a bit counterintuitive..

levitte replied:
Ah!  Good question, and I think you have the answer if you look in `providers/implementations/keymgmt/build.info`, see for example how the RSA keymgmt is included.
(it's not quite done the way I was imagining it, but that's a cleanup that can be done later)

romen replied:
Please have a look at 4dfad461e4a28a1b0e82c48f308c7dbba8ad3e3e

I had to decide if I should have a

```C
#if defined(FIPS_MODULE) && !defined(OPENSSL_NO_SM2)
# define OPENSSL_NO_SM2
#endif
```

section at the top of `providers/implementations/keymgmt/ec_kmgmt.c` or (like I ended up doing) adding a define for the libfips.a target.

What I am not sure about is: by setting

```
DEFINE[../../libfips.a]=OPENSSL_NO_SM2
```

am I overwriting the previous definition or adding to it?
If it is overwriting, is it limited to the sources defined in this `build.info` file or affects definitions at a global level?
Does the syntax support `+=` or something to that effect?

romen replied:
> It's actually strange that none other is needed that way

I was thinking myself about the reason, and I believe the difference is that the other SM2 specific functions are not referenced from `libimplementations.a` but only from code compiled as part of `libcrypto.a` for the default provider and from code in `libfips.a` which masks references to the SM2 calls using the `FIPS_MODULE` guard.


@slontis on a side note, I noticed the FIPS provider is including the ECX stuff, is that intentional?

levitte commented on providers/implementations/keymgmt/build.info:
```diff
@@ -14,7 +13,10 @@ IF[{- !$disabled{dsa} -}]
 IF[{- !$disabled{ec} -}]
-  SOURCE[$EC_GOAL]=ec_kmgmt.c
+  SOURCE[../../libfips.a]=ec_kmgmt.c
+  DEFINE[../../libfips.a]=OPENSSL_NO_SM2
```
You should consider `#ifndef FIPS_MODULE` in ec_kmgmt.c. That way, there's no need for this define

romen replied:
I did: https://github.com/openssl/openssl/pull/13359#discussion_r527063051

Do you think that would be better than using a define here?

romen replied:
I was playing with this as an alternative

```diff
diff --git a/providers/implementations/keymgmt/build.info b/providers/implementations/keymgmt/build.info
index b05501998f..f434a720bc 100644
--- a/providers/implementations/keymgmt/build.info
+++ b/providers/implementations/keymgmt/build.info
@@ -14,8 +14,6 @@ IF[{- !$disabled{dsa} -}]
 ENDIF
 IF[{- !$disabled{ec} -}]
   SOURCE[../../libfips.a]=ec_kmgmt.c
-  DEFINE[../../libfips.a]=OPENSSL_NO_SM2
-
   SOURCE[../../libnonfips.a]=ec_kmgmt.c
 ENDIF
 
diff --git a/providers/implementations/keymgmt/ec_kmgmt.c b/providers/implementations/keymgmt/ec_kmgmt.c
index cc88bb9278..e71c3344a6 100644
--- a/providers/implementations/keymgmt/ec_kmgmt.c
+++ b/providers/implementations/keymgmt/ec_kmgmt.c
@@ -13,6 +13,13 @@
  */
 #include "internal/deprecated.h"
 
+#if defined(FIPS_MODULE) && !defined(OPENSSL_NO_SM2)
+# define OPENSSL_NO_SM2
+#endif
+#if defined(FIPS_MODULE) && !defined(OPENSSL_NO_EC2M)
+# define OPENSSL_NO_EC2M
+#endif
+
 #include "e_os.h" /* strcasecmp */
 #include <string.h>
 #include <openssl/core_dispatch.h>
@@ -27,7 +34,6 @@
 #include "prov/providercommonerr.h"
 #include "prov/provider_ctx.h"
 #include "internal/param_build_set.h"
-#include "crypto/sm2.h"
 
 static OSSL_FUNC_keymgmt_new_fn ec_newdata;
 static OSSL_FUNC_keymgmt_gen_init_fn ec_gen_init;
@@ -50,7 +56,10 @@ static OSSL_FUNC_keymgmt_import_types_fn ec_import_types;
 static OSSL_FUNC_keymgmt_export_fn ec_export;
 static OSSL_FUNC_keymgmt_export_types_fn ec_export_types;
 static OSSL_FUNC_keymgmt_query_operation_name_fn ec_query_operation_name;
+
 #ifndef OPENSSL_NO_SM2
+# include "crypto/sm2.h"
+
 static OSSL_FUNC_keymgmt_gen_fn sm2_gen;
 static OSSL_FUNC_keymgmt_get_params_fn sm2_get_params;
 static OSSL_FUNC_keymgmt_gettable_params_fn sm2_gettable_params;
```

but somehow it felt better to define `OPENSSL_NO_SM2` (and `OPENSSL_NO_EC2M`, but maybe in a separate PR, as it is unrelated here) at the `build.info` rather than at the top of the file.
I don't have a strong preference either way.

levitte replied:
I would rather have something that is like this:

```
#ifndef FIPS_MODULE

/* ... SM2 stuff ... */

#endif
```

Where the SM2 stuff inside includes the `#ifndef OPENSSL_SM2` clauses.

levitte replied:
This is for detection reasons...

romen replied:
I am not particularly fun of the proliferation of ifndef (and nested ones) in the code, but I will comply.

I am just curious, could you please elaborate on the benefits of the multiple nested ifdefs vs my proposal of conditionally declaring OPENSSL_NO_whatever at the incipit of the file?
(really, this doesn't want to be polemic, I just desire to understand the motivation better)

levitte replied:
It's to make detection of what is FIPS and what isn't FIPS source.  We have a plan and I have a long overdue PR that will parse through source and make a checksum over all the FIPS source, and therefore need an easy enough way to skip over non-FIPS source.

This is the same problem as, for example, `#ifndef OPENSSL_NO_XXX` in headers to detect specific features, something `util/mknum.pl` does.  We shouldn't have to write a complete AI to do this stuff, and that requires a bit of compliance when writing the source.

romen replied:
Thanks for the explanation!

See 506418b62862a4cd5b90bf987ee11a203c525295 for the fixup.

romen commented:
I just pushed an empty commit (to be dropped) to trigger extended tests, just in case.

Apart from that, this is just waiting for OMC to start a vote about it.

romen commented:
The appveyor fault is weird, I would be surprised if it is related to this PR:

```
"cl"  /Zi /Fdossl_static.pdb /MT /Zl /Gs0 /GF /Gy /W3 /wd4090 /nologo /O2 -I "." -I "crypto" -I "include" -I "providers\implementations\include" -I "providers\common\include" -I ".." -I "..\crypto" -I "..\include" -I "..\providers\implementations\include" -I "..\providers\common\include" -D"L_ENDIAN" -D"OPENSSL_PIC" -D"OPENSSLDIR=\"C:\\Program Files\\Common Files\\SSL\"" -D"ENGINESDIR=\"C:\\Program Files\\OpenSSL\\lib\\engines-3\"" -D"MODULESDIR=\"C:\\Program Files\\OpenSSL\\lib\\ossl-modules\"" -D"OPENSSL_BUILDING_OPENSSL" -D"OPENSSL_SYS_WIN32" -D"WIN32_LEAN_AND_MEAN" -D"UNICODE" -D"_UNICODE" -D"_CRT_SECURE_NO_DEPRECATE" -D"_WINSOCK_DEPRECATED_NO_WARNINGS" -D"NDEBUG" -DOPENSSL_SMALL_FOOTPRINT  -c /Foproviders\common\libnonfips-lib-capabilities.obj "..\providers\common\capabilities.c"
capabilities.c
..\providers\common\capabilities.c(162): error C2059: syntax error: '}'
..\providers\common\capabilities.c(174): warning C4034: sizeof returns 0
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\bin\HostX64\x64\cl.EXE"' : return code '0x2'
Stop.
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\bin\HostX64\x64\nmake.exe"' : return code '0x2'
Stop.
```

I can't see the travis build!

romen commented:
### Further update on the vote issue

During the OTC meeting today, the attending OMC members suggested to have an OTC vote to qualify this kind of changes as bug fixes, which would circumvent the need for an explicit OMC approval for the behavior change.

Here is the vote proposal: https://www.mail-archive.com/openssl-project@openssl.org/msg02274.html

romen commented:
> ### Further update on the vote issue
> During the OTC meeting today, the attending OMC members suggested to have an OTC vote to qualify this kind of changes as bug fixes, which would circumvent the need for an explicit OMC approval for the behavior change.
> 
> Here is the vote proposal: https://www.mail-archive.com/openssl-project@openssl.org/msg02274.html

The OTC vote is now [closed and accepted](https://www.mail-archive.com/openssl-project@openssl.org/msg02331.html).

I'll now proceed to remove the OTC hold label and rebase on latest master

romen commented:
Force pushed, now it's time to wait for CI to run and reviewers to review!

romen commented on crypto/sm2/sm2_key.c:
```diff
@@ -38,7 +38,7 @@ int sm2_key_private_check(const EC_KEY *eckey)
     if (BN_cmp(priv_key, BN_value_one()) < 0
         || BN_cmp(priv_key, max) >= 0) {
-        ECerr(0, EC_R_INVALID_PRIVATE_KEY);
+        ERR_raise(ERR_LIB_SM2, EC_R_INVALID_PRIVATE_KEY);
```
@levitte is this fine? (using an `EC_R_`)
Or should I create a new dedicated `SM2_R_INVALID_PRIVATE_KEY`?

InfoHunter replied:
I think `SM2_R` is better...

romen replied:
Pushing fixup with the new error reason in 3, 2, 1 ...

romen commented:
3032f0c174 fixes a failure in `no-deprecated` builds detected by the CI.

As a sidenote, I noticed that `make cmd-nits` is currently failing in `no-deprecated` builds at the moment: is that intended?
~~~
doc/man1/openssl-list.pod: list does not implement -cipher-commands
doc/man1/openssl-list.pod: list does not implement -digest-commands
make: *** [Makefile:3675: cmd-nits] Error 1
~~~

t8m reviewed (approved):
LGTM

openssl-machine commented:
This pull request is ready to merge

romen commented:
Just merged to `master` with

- c5bc5ec849 *[test] Add `pkey -check` validation tests*
- ed37336b63 *[apps/pkey] Return error on failed `-[pub]check`*
- 4554988e58 *[test][pkey_check] Add invalid SM2 key test*
- 9e49aff2aa *Add SM2 private key range validation*
- 1330093b9c *[test][pkey_check] Add more invalid SM2 key tests*

I had to manually fix the conflict on the `CHANGES.md` file but it was trivial, given that it was intentionally left as a separate fixup commit.

Thanks everyone, and @bbbrumley for reporting the original issue.
